### DISCUSSION TITLE
app flavors and different providers

### DISCUSSION THREAD

**Discussion by mars3142** [rrousselGit/riverpod#2825] [Q&A]
I do have a question. Our app has three flavors (dev, prev and prod). With get_it I created 3 config classes which implements the base config class (which is a simple interface). After that I added the dependency depending on the flavor into the DI system. To retrieve the implementation, I just looked for the abstract class and get the correct implementation.

How would be the same behavior with Riverpod or should I do a complete different approach for this?

**Comment by ynlvko** [comment]
You can use `overrides` for the root `ProviderScope`.

**Reply by mars3142** [reply]
Hmm. Interesting idea. My current solution:

    final _configDevProvider = Provider<Config>((ref) {
      return ConfigDevelopment();
    });
    
    final _configPrevProvider = Provider<Config>((ref) {
      return ConfigPreview();
    });
    
    final _configProdProvider = Provider<Config>((ref) {
      return ConfigProduction();
    });

    class _EnvironmentNotifier extends StateNotifier<Environment> {
      _EnvironmentNotifier() : super(Environment.production);
    
      void setEnvironment(Environment environment) {
        state = environment;
      }
    }
    
    /// Current Environment/Flavor (dev, preview, production)
    final environmentProvider = StateNotifierProvider<_EnvironmentNotifier, Environment>((ref) => _EnvironmentNotifier());
    
    /// Get the Config based on the Environment/Flavor
    final configProvider = StateProvider<Config>((ref) {
      final environment = ref.watch(environmentProvider);
      switch (environment) {
        case Environment.preview:
          return ref.read(_configPrevProvider);
    
        case Environment.production:
          return ref.read(_configProdProvider);
    
        default:
          return ref.read(_configDevProvider);
      }
    });

And now I just need to set the environment in main with

    final container = ProviderContainer();
    container.read(environmentProvider.notifier).setEnvironment(environment);

and after that

    runApp(
      UncontrolledProviderScope(
        container: container,
        child: Home(),
      ),
    );

@ynlvko Your solution could be better, because of tree shaking it will remove the dev/prev config from the binary in production. Will check this later.
